### PR TITLE
Adds Aws S3 Adapter 

### PR DESCRIPTION
**Pull request by adidonato**
# TL;DR
Adds support for Amazon S3

## Synopsis
This extension allows for streaming files from the ETL app directly to aws s3 in a date partitioned fashion so that they can be automatically ingested in Hive / Spark (even databricks parquet delta tables - this is what I personally use it for). There is also a Makefile for easy building and pushing to `aws ecr`.  
There are no tests included in the PR but this has code been tested in production already at different Web3 companies :trollface: 

## Usage 
Simply use the `s3://` root path when passing the `--output` flag and the output will be automatically routed to s3.
## Requirements
`boto3` and AWS keys in the env (docker or not). 
### Example 
```bash
docker run --env AWS_ACCESS_KEY="" AWS_SECRET_KEY="" ethereum-etl:latest stream --start-block 12345678 --output s3://$your-aws-bucket --environment prod --chain ethereum
```

### CODE REVIEW

File: blockchainetl/jobs/exporters/s3_item_exporter.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from blockchainetl.file_utils import get_file_handle, close_silently
+from datetime import datetime
+import csv
+import json
+
+from uuid import uuid4
+
+class S3ItemExporter(CompositeItemExporter):
+
+    def __init__(self, filename_mapping=None, bucket=None, converters=(), environment='dev', chain='ethereum'):
```
Comment (medvedev1088) [inline blockchainetl/jobs/exporters/s3_item_exporter.py:16]: It seems the filename_mapping parameter is not used

File: blockchainetl/jobs/exporters/s3_item_exporter.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                reader = csv.DictReader(f, delimiter=",")
+                next(reader)
+                for row in reader:
+                    rows.append(row.get('item_timestamp'))
+            elif file.endswith('json'):
+                for i in f:
+                    data = json.loads(i)
+                    rows.append(data.get('item_timestamp'))
+
+            minimus = min(rows, key=lambda x: datetime.strptime(x.split('T')[0], '%Y-%m-%d'))
```
Comment (medvedev1088) [inline blockchainetl/jobs/exporters/s3_item_exporter.py:64]: It is possible that rows will contain blocks with different dates, and some data will end up in the wrong date partition?

**Comment by medvedev1088** [comment]
Thanks for the PR. I've added some comments

**Comment by adidonato** [comment]
Thanks and sorry for the late reply, I have been away.
Replied and will make some changes as pointed out

File: blockchainetl/jobs/exporters/s3_item_exporter.py (hunk 1 above)
Comment (adidonato) [reply blockchainetl/jobs/exporters/s3_item_exporter.py:16]: yes this is redundant, will clean it up

File: blockchainetl/jobs/exporters/s3_item_exporter.py (hunk 2 above)
Comment (adidonato) [reply blockchainetl/jobs/exporters/s3_item_exporter.py:64]: yes absolutely, this issue never happened to me as I was doing block by block i.e. batch size of 1. Here we need a check that splits the file according to the appropriate date partition (if the batch size is > 1), will follow up when I can!

**Comment by Menniti** [comment]
Hello Guys,

Very good feature @adidonato 
@medvedev1088, do you know when this will be merged to new release? I need exactly this feature!

Thank you

**Comment by zachary-newtonco** [comment]
I'm patiently awaiting this stream into aws s3 feature @medvedev1088 and @adidonato. Any idea when it can be merged? Very helpful feature indeed, thank you both!

**Comment by FahdW** [comment]
Curious, how does this work when you need to backfill the blocks before. Say you start from block 5M, do you backfill all the 1-4.9M will this clip and overwrite the CSV file or will it append it? Because my assumption is the sync will take the entirety of the file. Just want to understand the stream a bit more when dealing with blob storage

File: blockchainetl/jobs/exporters/s3_item_exporter.py (hunk 2 above)
Comment (LukvonStrom) [reply blockchainetl/jobs/exporters/s3_item_exporter.py:64]: @adidonato  Do you have a solution yet?

**Comment by FahdW** [comment]
I just reviewed everything, you are using DictReader and then it'll forward the progress of the row, and you will hit EOF or skip the first entry into your csv. And it seems to be lagging behind what it is actually doing.

**Review by namoona1 — COMMENTED** [review]
I am not sure where the issue is however when I use the S3 connector when uploading  to S3 it skips a block. Any help will be appreciated

**Comment by FahdW** [comment]
It actually doesn't, it will fail and then grab the block at the next pass. It is a little buggy tbh

**Comment by namoona1** [comment]
> It actually doesn't, it will fail and then grab the block at the next pass. It is a little buggy tbh

I am confused what do you mean by "next pass", I am not seeing any errors in outputs logs
